LogFile=<stdout> redirects log to stdout, which is useful for debugging

alexpevzner opened the pull request:
Hi!

This small pull requests adds possibility to redirect logging to the **stdout** instead of the log file, using the following simple syntax in the **xrdp.ini** file:
```
LogFile=-
```
This is useful for xrdp debugging and doesn't affect normal operations

alexpevzner committed:
LogFile=- redirects log to stdout, which is useful for debugging

metalefty commented on common/log.c:
```diff
@@ -61,8 +62,15 @@ internal_log_file_open(const char *fname)
     {
-        ret = open(fname, O_WRONLY | O_CREAT | O_APPEND | O_SYNC,
-                   S_IRUSR | S_IWUSR);
+	if (g_strcmp(fname, "-") != 0)
```
`-` is a common symbol to represent standard out in command line arguments. However, this looks a little bit odd because the context is not a command line.

What about using a word "stdout" instead of `-`.

```
Logfile=stdout
```

alexpevzner replied:
`stdout` looks like legitimate file name, an this is very unlikely that somebody will want to name log file with a single dash (`-`) name. This is why I've chosen this name.

If you fill it looks a little bit odd, what about `<stdout>`, with angle braces around the word?

And in any case, I consider it as a debug mode, not intended for the common user.

metalefty replied:
I personally like `<stdout>`.

metalefty commented on docs/man/xrdp.ini.5.in:
```diff
@@ -230,7 +230,7 @@ The following parameters can be used in the \fB[Logging]\fR section:
 .TP
 \fBLogFile\fR=\fI@localstatedir@/log/xrdp.log\fR
-This options contains the path to logfile. It can be either absolute or relative.\fR
+This options contains the path to logfile. It can be either absolute or relative. If set to "-", log will go to stdout, which is useful for debugging\fR
```
I think the same fix should be applied to sesman.ini.

alexpevzner commented:
OK. Let's agree on that: `<stdout>` and implement it in the `sesman.ini` too. I'll come a bit later with the updated PR...

alexpevzner committed:
Use LogFile=<stdiut> instead of LogFile=- to redirect log to stdout

matt335672 commented on docs/man/xrdp.ini.5.in:
```diff
@@ -230,7 +230,7 @@ The following parameters can be used in the \fB[Logging]\fR section:
 .TP
 \fBLogFile\fR=\fI@localstatedir@/log/xrdp.log\fR
-This options contains the path to logfile. It can be either absolute or relative. If set to "-", log will go to stdout, which is useful for debugging\fR
+This options contains the path to logfile. It can be either absolute or relative. If set to \fB<stdout>\fR, log will go to stdout, which is useful for debugging\fR
```
Pardon me for jumping in, but there's a (very) small issue here which hasn't been discussed.

At the moment, the log file is created with `O_APPEND` and `O_SYNC`. This has the useful effect of making sure that if an application forks and continues logging, the log messages are interleaved as you'd expect. There's a downside to this in that it really hits performance but that's a separate issue.

If we're logging to stdout, we lose the benefits of the interleaving, and it's possible with buffering for the main processes and the sub-processes to trample over each other's messages. If we're debugging, that's fine, but it's not something we should be using in production.

Consequently, in the above message I'd like to see *, which is useful for debugging* replaced with *. Use for debugging only.*

Does that make sense?

alexpevzner replied:
Hi @metalefty

I've updated the PR according to to your comments

matt335672 replied:
Comment update above?

metalefty replied:
I'm fine with that if you reflect Matt's comment.

alexpevzner replied:
Hi,

sorry for delay. I've updated the documentation, as @matt335672 suggests

alexpevzner committed:
Code formatting fixed

alexpevzner committed:
LogFile in sesman.ini(5) updated too
docs/man/sesman.ini.5.in

alexpevzner commented:
Hi @matt335672,

I'm not pretty sure this mode is only useful for debugging, so I'd prefer to avoid such a strong wording. But if others agree that this is debug-only mode, I have no problems to update man pages. My personal interest to this mode is debugging.

matt335672 commented:
Hi @alexpevzner 

The reason for the wording is that (as I mentioned above) stdout logging is buffered and doesn't go through the locking mechanism. The advantage of this is that logging is really quick. The disadvantage is that if you're running multiple sub-processes, the lines from all of these will not necessarily be in the expected order and also may interrupt each other halfway through.

When debugging. logging speed is important and you're only likely to have a very small number of sub-processes. In production you are more likely to have more sub-processes, and so when things do go wrong it's important to have an order and for no interleaving to have taken place.

That's just my opinion of course, but I do a fair bit of xrdp debugging myself!. Feel free to come back at me if you disagree with my analysis.

Commit:
Documentation of log file redirection to <stdout> updated

Since multiple processes can write logs simultaneously, and there
is no explicit synchronization between them, logs redirected to
stdout may interleave. When logging to the disk file, synchronization
is obtained by using a combination of O_APPEND and O_SYNC flags when
log file is being opened, which is not possible when <stdout>
redirection is in use.

So documentation is updated to cleanly notify that this is debug-only
feature, not for regular use

matt335672 commented:
I'm merging this now as @metalefty has indicated he's happy with it.

alexpevzner commented:
Thanks!

matt335672 commented:
Not at all - thank you!
